vm/src/execute.rs

```diff
@@ -70,14 +69,25 @@ pub fn execute(
     // with gas limit smaller than deploy charge.
     deploy_check(tx, config)?;
 
+    if config.with_public_sender {
+        let _ = session
+            .set_meta(Metadata::PUBLIC_SENDER, tx.moonlight_sender().copied());
+    }
+
     // Spend the inputs and execute the call. If this errors the transaction is
     // unspendable.
-    let mut receipt = session.call::<_, Result<Vec<u8>, ContractError>>(
+    let receipt = session.call::<_, Result<Vec<u8>, ContractError>>(
         TRANSFER_CONTRACT,
         "spend_and_execute",
         tx.strip_off_bytecode().as_ref().unwrap_or(tx),
         tx.gas_limit(),
-    )?;
+    );
+
+    if config.with_public_sender {
+        let _ = session.remove_meta(Metadata::PUBLIC_SENDER);
+    }
```

Comment:
What is the reason of removing the sender again after calling spend_and_execute?

Reply:
in theory is not needed, because it will be overwritten by the next transaction.

However, the session is still open for further operation (like reward and deploy) so I'm not sure

ps: on second tough, there is a bug in that code. removing the sender before the deploy can cause issue if the "init" method require that ABI

Reply:
Indeed, that was my line of thought: why remove it right after `spend_and_execute` when we possibly still need it in the rest of the `execute` function.
Maybe even inject the public_sender metadata in `rusk` before calling `execute` (and removing it afterwards again). That way the execute config might not even need the `with_public_sender` variable.

Reply:
We need a config to know at which height inject it to preserve backward compatibility

Having (or not having) the value in the metadata can change the result of the execution. Meaning that a sync from genesis could result in different state root
Ps: this can happen only if there is a smart contract deployed that use that ABI before it's activated

Reply:
I don't understand. If in rusk we inject the public-sender only when we are at the specified block-height, the ABI simply wouldn't be available if a contract tries to use it at an earlier block-height (the call would panic).

Reply:
I misunderstood my bad
I thought you were suggesting to alway inject, instead you were suggesting to inject it from outside

Anyway, injecting it from rusk would be an option too, but it's against the logic of having the vm::execute method self/consistent (you could not test it without rusk)

Anyway, i would try to go with this and, if possibile, remove the conditional injection if not needed (we'll only know after the activation)

Reply:
`execute` to be self-contained is a good argument!